Adding 3D-Plot #200

koaning opened the pull request:
This adresses https://github.com/RasaHQ/whatlies/issues/200. 

```python
emb.transform(Pca(3)).plot_3d(annot=True)
```

![image](https://user-images.githubusercontent.com/1019791/92602254-34114600-f2ae-11ea-8028-4ba920e96dde.png)

```python
emb.transform(Pca(3)).plot_3d("king", "dude", "green", annot=True)
```

![image](https://user-images.githubusercontent.com/1019791/92602270-383d6380-f2ae-11ea-8eff-703d3cd4904b.png)

There's a few things about this PR worth discussing. 

1. The color property works differently in 3d charts. It seems as if the color is automatically altered depending on the depth of the point. Since this is a sensible way of dealing with it I'm wondering if this is something we want to influence directly. 
2. I'm currently adding an absolute value to offset the text from the 3d point. It kind of works but it depends on the zoom level. @mkaze might this be better as an argument? 
3. It feels like we might be able to generalise the plots. They always have a phase where a DataFrame is created for the plot and it seems like a lot of code is repeating itself. This is not something I want to adress in this PR but it is something I've observed. 
4. Testing 3D plots is different. When you call `ax.collections[0].get_offsets()` then you seem to get the 2d coordinates of the points that are plotted instead of the original 3d ones. I've added some basic tests for now, but I'll gladly hear it if we need more. I'm still investigating how the 3d properties work exactly in matplotlib.

mkaze commented on whatlies/embeddingset.py:
```diff
@@ -679,6 +679,119 @@ def plot(
+                returns a scalar value as output. To set different metrics for different axes, a list or a tuple of
+                the same length as `axes` could be given. By default (`None`), normalized scalar projection (i.e. `>` operator) is used.
+            annot: drawn points should be annotated
+
```
What about adding `x_label`, `y_label`, `z_label` and `title` arguments? For the sake of consistency!

mkaze commented on whatlies/embedding.py:
```diff
@@ -299,7 +299,11 @@ def _get_plot_axis_metric_callable(axis_metric) -> Optional[Callable]:
-            if axis_metric == "cosine_similarity":
+            if axis_metric == "cosine":
+                raise ValueError(
+                    "Please be specific, do you want `cosine_distance` or `cosine_similarity`."
```
Since it's a question, a question mark should be put at the end instead.

mkaze commented on whatlies/embeddingset.py:
```diff
@@ -679,6 +679,119 @@ def plot(
+            axes_metric: the metric used to project each embedding on the axes; only used when the corresponding
+                axis is a string or an `Embedding` instance. It could be a string (`'cosine_similarity'`,
+                `'cosine_distance'` or `'euclidean'`), or a callable that takes two vectors as input and
+                returns a scalar value as output. To set different metrics for different axes, a list or a tuple of
+                the same length as `axes` could be given. By default (`None`), normalized scalar projection (i.e. `>` operator) is used.
```
"To set different metrics for each of the three axes, a list or tuple of three elements could be given."

The current one is the one used for `plot_interactive_matrix`.

mkaze commented on whatlies/embeddingset.py:
```diff
@@ -679,6 +679,119 @@ def plot(
+        y_axis: Union[int, str, Embedding] = 1,
+        z_axis: Union[int, str, Embedding] = 2,
+        color: str = None,
+        axes_metric: Optional[Union[str, Callable, Sequence]] = None,
```
To make it consistent with `plot` and `plot_interactive` let's use `axis_metric` instead of `axes_metric`.
I used `axes_metric` in `plot_interactive_matrix` because there was a `axes` argument. Maybe I shouldn't have done that!

mkaze commented:
> 1. The color property works differently in 3d charts. It seems as if the color is automatically altered depending on the depth of the point. Since this is a sensible way of dealing with it I'm wondering if this is something we want to influence directly.

Well, actually this could be controlled by `depthshade` argument of `scatter3D` which is `True` by default (see [documentation](https://matplotlib.org/3.1.1/api/_as_gen/mpl_toolkits.mplot3d.axes3d.Axes3D.html?highlight=scatter3d#mpl_toolkits.mplot3d.axes3d.Axes3D.scatter3D)); so this is something that could be controlled. Actually, in order to not provide a lot of custom arguments and at the same time give more flexibility to the user, I am thinking of adding a `**kwargs` argument for all plot methods where the user can provide additional arguments which could be potentially passed to the main underlying plot method (e.g. in this case it would be `scatter3D`) and we can clarify this in the docstrings. If you see the documentation, you will see that this is something that matplotlib does as well. However, I am not suggesting that this should be done in this PR.

> 2. I'm currently adding an absolute value to offset the text from the 3d point. It kind of works but it depends on the zoom level. @mkaze might this be better as an argument?

Well, I think `0.2` is a bit high, especially if the coordinate values is in the order of less than 1. `0.05` might be a better option. But, I am not sure if it's a good idea to control this using a separate argument.  

> 3. It feels like we might be able to generalise the plots. They always have a phase where a DataFrame is created for the plot and it seems like a lot of code is repeating itself. This is not something I want to adress in this PR but it is something I've observed.

That's right. There is a data preparation stage with some common steps in most of the plot methods. There is a potential to do some refactoring there (even for the methods that currently use `handle_2d_plot`). Though, this also could be addressed when refactoring the visualization API (#202) as well.

> 4. Testing 3D plots is different. When you call `ax.collections[0].get_offsets()` then you seem to get the 2d coordinates of the points that are plotted instead of the original 3d ones. I've added some basic tests for now, but I'll gladly hear it if we need more. I'm still investigating how the 3d properties work exactly in matplotlib.

First, I thought there must be some 3D to 2D projection is happening. So I looked in the methods list and found the `get_offset_transform`. However, although it seems to have some kind of projection matrix but none of its transform method worked as I expected. 
Next, I searched a bit and found [this answer](https://stackoverflow.com/a/51716781/2099607) on SO. This is basically what you want, however in the current implementation of `plot_3d` the return value of `scatter3D` is not returned (instead axis value is returned which seems to be a bit different). So if we are interested in automated testing of this plot, either the return value should be changed or we should continue searching for a workaround. My opinion is that to use the return value of `Scatter3D` as the return value of the method because then `.axes` property of that value would also gives us the `plt.axes(...)` return value.

mkaze commented on whatlies/embeddingset.py:
```diff
@@ -746,23 +755,23 @@ def plot_3d(
             x_axis_metric = Embedding._get_plot_axis_metric_callable(x_axis_metric)
             x_val = self.compare_against(x_axis, mapping=x_axis_metric)
-            x_lab = x_axis.name
+            x_lab = x_label if x_label else x_axis.name
```
This line should be put after the `if/else`, otherwise it won't be effective when `if` condition is True. Further it should be like `x_lab = x_label if x_label is not None else ....`, otherwise when `x_label = ''` this would not be effective (someone may want to not use a label for an axis using this trick). The same goes for labels of other axes.

mkaze commented on whatlies/embeddingset.py:
```diff
@@ -698,12 +702,17 @@ def plot_3d(
-                returns a scalar value as output. To set different metrics for different axes, a list or a tuple of
-                the same length as `axes` could be given. By default (`None`), normalized scalar projection (i.e. `>` operator) is used.
+                returns a scalar value as output. To set different metrics of the three different axes,
+                a list or a tuple of the same length as `axes` could be given. By default (`None`),
```
As I mentioned in my previous comment, there is no "\`axes\`" for this method!

koaning commented:
@mkaze I've made some of the changes. 

- I've added `title` and `x/y/z_label` arguments. 
- I'm using the `ax.collections[0]._offsets3d` to test some of the drawn points. 
- Renamed some `axes`-arguments to `axis`-arguments. I think I've got all of them now. 
- I've added support for the `color` property which wasn't in there before. I'm not touching the depth, just allowing the user to pass a property like in the other methods for consistency. 

The title looks a bit ... strange in a 3d plot. But for consistency it's been added. 

```python
plt.figure(figsize=(7,7))
embset = emb.add_property("group", lambda d: random.random()).transform(Pca(3))
ax = embset.plot_3d(annot=True, title="foobar", color='group')
```

![image](https://user-images.githubusercontent.com/1019791/92708141-2e6f3b00-f356-11ea-85df-ddf34c03fecd.png)

```python
plt.figure(figsize=(7,7))
embset = emb.add_property("group", lambda d: random.random() > 0.5).transform(Pca(3))
embset.plot_3d(annot=True, title="foobar", color='group')
```

![image](https://user-images.githubusercontent.com/1019791/92708166-329b5880-f356-11ea-970d-638480a8c483.png)


My only caveat in using this method right now is the manual figsize that I am setting. With setting it, this is what the last chart looks like; 

![image](https://user-images.githubusercontent.com/1019791/92708278-4941af80-f356-11ea-839d-7465384242c6.png)

The shape is a bit clumped together and you can also see that there is some part of the text missing on the right.

mkaze commented:
@koaning All right! Thanks for the changes.
Just one note about the `color` argument: how are we supposed to assign a single color to the points without using a custom property? For example, `plot_3d(...., color='red')`? I am asking this because the behavior of `color` argument in Altair plot methods is also the same.

koaning commented:
So currently if you pass `color='red'` to `.plot_interactive` then it kind of gets ignored. 

<img width="659" alt="image" src="https://user-images.githubusercontent.com/1019791/92738176-04c90a80-f37c-11ea-9534-a0dcc342a471.png">

A similar thing happens in `plot_3d`. 

<img width="590" alt="image" src="https://user-images.githubusercontent.com/1019791/92738574-5a9db280-f37c-11ea-826d-4e3187a4a3c7.png">

Effectively there's an internal column column with only `""` values that are made as a placeholder and it will be using this column to assign the color using what-ever the plotting backend is. I'm wondering what the better behavior is in this scenario. 

1. We might want to take the matplotlib standard and allow for a subset of strings that represent colors to change the colors of all the points. This is a standard in matplotlib, but other libraries deal with this differently.
2. We might throw an error instead, indicating that you can only pass a reference to a column name with the `color` argument. It won't be flexible, but might be easier to be consistent across different types of plots. 

We might also defer this decision to another PR since it this topic spans not just the `.plot3d()` method.

koaning commented:
Other than that, this good to merge @mkaze?

mkaze reviewed (approved):
It looks good to me. Thanks!
